mempool/cat/pool.go

```diff
@@ -0,0 +1,786 @@
+	seenByPeersSet *SeenTxSet
+
+	// Store of wrapped transactions
+	store *store
+
+	// broadcastCh is an unbuffered channel of new transactions that need to
+	// be broadcasted to peers. Only populated if `broadcast` is enabled
+	broadcastCh      chan types.TxKey
+	broadcastMtx     sync.Mutex
+	txsToBeBroadcast map[types.TxKey]struct{}
+}
+
+// NewTxPool constructs a new, empty content addressable txpool at the specified
+// initial height and using the given config and options.
+func NewTxPool(
+	logger log.Logger,
+	cfg *config.MempoolConfig,
+	proxyAppConn proxy.AppConnMempool,
+	height int64,
+	options ...TxPoolOption,
+) *TxPool {
+	txmp := &TxPool{
+		logger:           logger,
+		config:           cfg,
+		proxyAppConn:     proxyAppConn,
+		metrics:          mempool.NopMetrics(),
+		rejectedTxCache:  NewLRUTxCache(cfg.CacheSize),
+		evictedTxs:       NewEvictedTxCache(evictedTxCacheSize),
+		seenByPeersSet:   NewSeenTxSet(),
+		height:           height,
+		preCheckFn:       func(_ types.Tx) error { return nil },
+		postCheckFn:      func(_ types.Tx, _ *abci.ResponseCheckTx) error { return nil },
+		store:            newStore(),
+		broadcastCh:      make(chan types.TxKey, 1),
+		txsToBeBroadcast: make(map[types.TxKey]struct{}),
+	}
+
+	for _, opt := range options {
+		opt(txmp)
+	}
+
+	return txmp
+}
+
+// WithPreCheck sets a filter for the mempool to reject a transaction if f(tx)
+// returns an error. This is executed before CheckTx. It only applies to the
+// first created block. After that, Update() overwrites the existing value.
+func WithPreCheck(f mempool.PreCheckFunc) TxPoolOption {
+	return func(txmp *TxPool) { txmp.preCheckFn = f }
+}
+
+// WithPostCheck sets a filter for the mempool to reject a transaction if
+// f(tx, resp) returns an error. This is executed after CheckTx. It only applies
+// to the first created block. After that, Update overwrites the existing value.
+func WithPostCheck(f mempool.PostCheckFunc) TxPoolOption {
+	return func(txmp *TxPool) { txmp.postCheckFn = f }
+}
+
+// WithMetrics sets the mempool's metrics collector.
+func WithMetrics(metrics *mempool.Metrics) TxPoolOption {
+	return func(txmp *TxPool) { txmp.metrics = metrics }
+}
+
+// Lock is a noop as ABCI calls are serialized
+func (txmp *TxPool) Lock() {}
+
+// Unlock is a noop as ABCI calls are serialized
+func (txmp *TxPool) Unlock() {}
+
+// Size returns the number of valid transactions in the mempool. It is
+// thread-safe.
+func (txmp *TxPool) Size() int { return txmp.store.size() }
+
+// SizeBytes return the total sum in bytes of all the valid transactions in the
+// mempool. It is thread-safe.
+func (txmp *TxPool) SizeBytes() int64 { return txmp.store.totalBytes() }
+
+// FlushAppConn executes FlushSync on the mempool's proxyAppConn.
+//
+// The caller must hold an exclusive mempool lock (by calling txmp.Lock) before
+// calling FlushAppConn.
+func (txmp *TxPool) FlushAppConn() error {
```

Comment:
Wondering why not acquiring the lock inside the function instead of delegating it to the caller?

Reply:
I don't know. 

The way I've been taught, having `Unlock` and `Lock` as exposed functions is bad practice. A mutex lock should be at the level of the component controlling concurrency. There's a lot of this design that I don't agree with but I don't want to change to avoid the size of the work that this is. Most of this code is copied across if you look at the priority and fifo mempool you will see similar patterns. 

In regards to this, flushing isn't even something that celestia cares about since the consensus and the application are within the same function so ABCI calls are just go method calls.

Reply:
> There's a lot of this design that I don't agree with but I don't want to change to avoid the size of the work that this is. 

Fair point!